torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1424,6 +1433,458 @@ def rbinary_magic_impl(self, other):
+                assertion=_boolref(denominator != 0),
+            )
+
+        @staticmethod
+        def floor(number: z3.ArithRef) -> Z3Node:
+            return Z3Node(Z3Ops.to_int(number).expr)
+
+        @staticmethod
+        def floordiv(numerator: z3.ArithRef, denominator: z3.ArithRef) -> Z3Node:
+            div = Z3Ops.div(numerator, denominator)
+            return Z3Node(Z3Ops.to_int(div.arith).expr, div.assertion)
+
+        @staticmethod
+        def ceil(number: z3.ArithRef) -> Z3Node:
+            ite = z3.If(Z3Ops.floor(number).arith < number, Z3Ops.floor(number + 1).arith, number)
+            return Z3Node(cast(z3.ArithRef, ite))
+
+        @staticmethod
+        def mod(p: z3.ArithRef, q: z3.ArithRef) -> Z3Node:
+            q = Z3Ops.to_int(q).arith
+            return Z3Node(
+                expr=Z3Ops.to_int(p).arith % q,
+                assertion=_boolref(q != 0),
+            )
+
+        @staticmethod
+        def pow(base: z3.ArithRef, exp: z3.ArithRef) -> Z3Node:
+            return Z3Node(
+                expr=base ** exp,
+                assertion=_boolref(z3.Or(base != 0, exp > 0)),
+            )
+
+        @staticmethod
+        def sqrt(number: z3.ArithRef) -> Z3Node:
+            # Square-root:
+            # 1. Only work with reals
+            # 2. The number should be positive or zero.
+            #    Otherwise, Z3 returns 'unknown'.
+            number = Z3Ops.to_real(number).arith
+            return Z3Node(
+                expr=number ** 0.5,
+                assertion=number >= 0
+            )
+
+    # Lifts a callable to be used in Z3.
+    #
+    # This function replaces the given 'op' by a function that:
+    #
+    #   1. Lifts the arguments into Z3 (i.e. make them inhabitants of Z3)
+    #
+    #   2. Calls an operation that corresponds to 'op', but works with Z3
+    #      inhabitants (left as is if it works as is)
+    #
+    #   3. Lifts the return statement into 'Z3Node'
+    def _operator_to_z3(op: Callable) -> Callable:
+        # Operations that have booleans as their argument.
+        # This is neeeded because the argument of some FX nodes were
+        # literal integers, instead of booleans. So, whenever this flag
+        # is set, we also convert ints to booleans.
+        boolean_ops = {operator.not_, operator.and_, operator.or_}
+        as_bool = op in boolean_ops
+
+        # Lifts the function into 'Z3Node' domain.
+        def lift(func):
+            def wrap(a) -> Z3Node:
+                if isinstance(a, Z3Node):
+                    return a
+                if isinstance(a, (z3.ArithRef, z3.BoolRef)):
+                    return Z3Node(a)
+                # Convert it into a Z3 value, if it is some of the supported
+                # types below. Finally, wrap it into a 'Z3Node'.
+                if isinstance(a, bool) or (as_bool and isinstance(a, int)):
+                    return Z3Node(z3.BoolVal(bool(a)))
+                if isinstance(a, (int, sympy.Integer)):
```

Comment:
it's not clear to me when you could ever get a sympy.Integer here

Reply:
Inside the function `evaluate_expr`, the variable `concrete_val` might be a `sympy.Integer`. So, we may end up creating an FX node like `operator.eq(expr, sympy.Integer(X))`.